### PR TITLE
Django migrations: generate on the fly via import hooks

### LINKED ISSUE #323: Django migrations: generate on the fly via import hooks
Currently, Django migrations are generated once via makemigrations, because I believed they had to be stored on disk, and generation at release time was complicated and generation at migration time was likely to fail if the process that runs migration can’t write to the disk.

@MarkusH came up with a very interesting idea:
Actually there is one abstraction layer between Django migrations and reading the disk that we can tap into: django uses the Python import mechanizm that exposes all kinds of hooks. If we can generate the migration modules in memory and convince Python that our migrations top-level module contains our in-memory modules, then we’ll be much better off !

Time to experiment !

### PR DESCRIPTION
**Pull request by ewjoachim**
Closes #323

### Successful PR Checklist:

- [x] Tests
  - [ ] (not applicable?)
- [x] Documentation < **A hell lot of it !**
  - [ ] (not applicable?)
- [x] Had a good time contributing?

### CODE REVIEW

File: tests/unit/contrib/django/test_migrations_magic.py [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+@pytest.fixture
+def clean_paths(mocker):
+    mocker.patch.object(sys, "meta_path", [])
+    mocker.patch.object(sys, "path_hooks", [])
+
+
+def test_load(clean_paths):
+    migrations_magic.load()
+    assert isinstance(
+        sys.meta_path[-1], migrations_magic.ProcrastinateMigrationsImporter
```
Comment (ewjoachim) [inline tests/unit/contrib/django/test_migrations_magic.py:166]: -1 is not needed anymore

File: tox.ini [hunk 2]
```diff
@@ -20,15 +20,12 @@ extras =
     lint
     django
```
Comment (ewjoachim) [inline tox.ini:21]: remove this line too

File: procrastinate/contrib/django/README.md [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+for conciseness.
+
+The end goal is that when we run `./manage.py migrate`, Django sees one Django migration
+for each sql migration in our `procrastinate/sql/migrations` folder. This means we want
+to "trick" Django into thinking that there are migrations in the `p.c.d.migrations`
+package, but this package doesn't exist on the disk.
+
+This all happens in  `migrations_magic` if you want to read the code along.
+
+Django introspects it's apps looking for migrations (in django.db.migrations.loader). It
```
Comment (elemoine) [inline procrastinate/contrib/django/README.md:21]:
```suggestion
Django introspects its apps looking for migrations (in django.db.migrations.loader). It
```

File: procrastinate/contrib/django/README.md [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+Python's import mechanisms are modular, there's multiple ways we can add some code for
+our use case. While not necessary for the following, reading Python's import
+documentation will help: https://docs.python.org/3/reference/import.html
+
+We could imagine doing step A by going either way: either `p.c.d.migrations` could exist
+and be empty or we could trick Python's import system into believing it exists. The
+first alternative seems simpler, but if we do that, in the B step,
+`pkgutil.iter_modules` will try to load submodules the same way this module was loaded:
+by reading the disk. So in the end, given we'll be tricking the import system anyway,
+we're doing it for both `p.c.d.migrations` and `p.c.d.migrations`.*
```
Comment (elemoine) [inline procrastinate/contrib/django/README.md:41]:
```suggestion
we're doing it for both `p.c.d.migrations` and `p.c.d.migrations.*`
```

?

(I am not sure.)

**Review by elemoine — APPROVED** [review]
Looks good to me, with suggestions. Also, remember you have your own comments to take into account before merging.

**Pull request merged**